node/v2/spy.js

```diff
@@ -0,0 +1,127 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var TypedError = require('error/typed');
+var inherits = require('util').inherits;
+var Transform = require('stream').Transform;
+var hex = require('hexer');
+
+var BrokenReaderStateError = TypedError({
+    type: 'tchannel.broken-reader-state',
+    message: 'reader in invalid state {state}',
+    state: null
+});
+
+module.exports = ChunkSpy;
+
+var States = {
+    PendingLength: 0,
+    Seeking: 1
+};
+
+function ChunkSpy(sink, options) {
+    if (!(this instanceof ChunkSpy)) {
+        return new ChunkSpy(sink, options);
+    }
+    options = options || {};
+    Transform.call(this, options);
+    var self = this;
+    self.hex = hex.Transform(options);
+    self.frameLengthSize = options.frameLengthSize || 2;
+    self.buffer = Buffer(0);
+    self.expecting = self.frameLengthSize;
+    self.state = States.PendingLength;
+    self.hex.pipe(sink);
+    switch (self.frameLengthSize) {
+        case 1:
+            self._readLength = Buffer.prototype.readUInt8;
+            break;
+        case 2:
+            self._readLength = Buffer.prototype.readUInt16BE;
+            break;
+        case 4:
+            self._readLength = Buffer.prototype.readUInt32BE;
+            break;
+        default:
+            throw new Error('unsupported frame length size');
+    }
+}
+
+inherits(ChunkSpy, Transform);
+
+ChunkSpy.prototype._transform = function _transform(chunk, encoding, callback) {
+    var self = this;
+    if (!callback) {
+        callback = emitIt;
+    }
+    if (self.buffer.length) {
+        self.buffer = Buffer.concat([self.buffer, chunk], self.buffer.length + chunk.length);
+    } else {
+        self.buffer = chunk;
+    }
+    while (self.buffer.length) {
+        switch (self.state) {
+            case States.PendingLength:
+                if (self.buffer.length >= self.expecting) {
+                    self.expecting = self._readLength.call(self.buffer, 0);
+                    self.state = States.Seeking;
+                    self.hex.reset();
```

Comment:
see this logic right here? it's where we reset hex dup session between each chunk boundary